File: C#/10. Regular Expression Matching.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        int n = s.Length;
        int m = p.Length;
        bool[,] dp = new bool[n+1,m+1];
        dp[0,0] = true;
        for(int i=0; i<=n; i++){
            for(int j=1; j<=m; j++){
                if(p[j-1] == '*'){
                    dp[i,j] = dp[i,j-2] || (i != 0 && (s[i-1] == p[j-2] || p[j-2] == '.') && dp[i-1,j]);
                }
                else if(i > 0 && (s[i-1] == p[j-1] || p[j-1] == '.')){
                    dp[i,j]=dp[i-1,j-1];
                }
            }
        }
        return dp[n,m];
    }
}

File: Java/10. Regular Expression Matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        int n = s.length();
        int m = p.length();
        boolean[][] dp = new boolean[n+1][m+1];
        dp[0][0] = true;
        for(int i=0; i<=n; i++){
            for(int j=1; j<=m; j++){
                if(p.charAt(j-1) == '*'){
                    dp[i][j] = dp[i][j-2] || (i != 0 && (s.charAt(i-1) == p.charAt(j-2) || p.charAt(j-2) == '.') && dp[i-1][j]);
                }
                else if(i > 0 && (s.charAt(i-1) == p.charAt(j-1) || p.charAt(j-1) == '.')){
                    dp[i][j]=dp[i-1][j-1];
                }
            }
        }
        return dp[n][m];
    }
}
